File: src/utils/dateUtils.ts

/**
 * Utility functions for standardized date handling
 * All dates in the system should use ISO 8601 format: YYYY-MM-DDTHH:MM:SS
 */

export const formatDateTimeForDatabase = (date: Date): string => {
  return date.toISOString().slice(0, 19); // Remove milliseconds and timezone
};

export const formatDateForDatabase = (date: Date): string => {
  return date.toISOString().slice(0, 10); // YYYY-MM-DD format
};

export const parseDatabaseDateTime = (dateString: string): Date => {
  // Handle both ISO format and SQLite datetime format
  if (!dateString) {
    throw new Error('Invalid date string');
  }
  
  // Ensure the date string is in proper ISO format
  let isoString = dateString;
  if (!dateString.includes('T')) {
    isoString = dateString.replace(' ', 'T');
  }
  
  // Remove timezone suffix if present for local time handling
  isoString = isoString.replace('Z', '').replace(/[+-]\d{2}:\d{2}$/, '');
  
  return new Date(isoString);
};

export const formatDisplayDateTime = (dateString: string, locale: string = 'ru-RU'): string => {
  try {
    const date = parseDatabaseDateTime(dateString);
    return date.toLocaleString(locale, {
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
      hour: '2-digit',
      minute: '2-digit'
    });
  } catch {
    return dateString; // Fallback to original string if parsing fails
  }
};

export const formatDisplayDate = (dateString: string, locale: string = 'ru-RU'): string => {
  try {
    const date = parseDatabaseDateTime(dateString);
    return date.toLocaleDateString(locale, {
      year: 'numeric',
      month: '2-digit',
      day: '2-digit'
    });
  } catch {
    return dateString; // Fallback to original string if parsing fails
  }
};

export const formatDisplayTime = (dateString: string, locale: string = 'ru-RU'): string => {
  try {
    const date = parseDatabaseDateTime(dateString);
    return date.toLocaleTimeString(locale, {
      hour: '2-digit',
      minute: '2-digit'
    });
  } catch {
    return dateString; // Fallback to original string if parsing fails
  }
};

export const getCurrentDateTime = (): string => {
  return formatDateTimeForDatabase(new Date());
};

export const getCurrentDate = (): string => {
  return formatDateForDatabase(new Date());
};

export const isToday = (dateString: string): boolean => {
  try {
    const date = parseDatabaseDateTime(dateString);
    const today = new Date();
    return date.toDateString() === today.toDateString();
  } catch {
    return false;
  }
};

export const calculateDuration = (startDateTime: string, endDateTime: string): number => {
  try {
    const start = parseDatabaseDateTime(startDateTime);
    const end = parseDatabaseDateTime(endDateTime);
    return Math.round((end.getTime() - start.getTime()) / (1000 * 60)); // Minutes
  } catch {
    return 0;
  }
};

export const addMinutesToDateTime = (dateString: string, minutes: number): string => {
  try {
    const date = parseDatabaseDateTime(dateString);
    const newDate = new Date(date.getTime() + minutes * 60000);
    return formatDateTimeForDatabase(newDate);
  } catch {
    return dateString;
  }
};

export const getCooldownEnd = (minutes: number = 30): string => {
  const futureTime = new Date(Date.now() + minutes * 60000);
  return formatDateTimeForDatabase(futureTime);
};

File: utils/date_utils.py

"""
Utility functions for standardized date handling in backend
All dates in the system use ISO 8601 format: YYYY-MM-DDTHH:MM:SS
"""

from datetime import datetime, timedelta
import sqlite3

def get_current_datetime() -> str:
    """Get current datetime in ISO format"""
    return datetime.now().isoformat()

def get_current_date() -> str:
    """Get current date in ISO format (YYYY-MM-DD)"""
    return datetime.now().date().isoformat()

def format_datetime_for_db(dt: datetime) -> str:
    """Format datetime object for database storage"""
    return dt.isoformat()

def parse_datetime_from_db(date_string: str) -> datetime:
    """Parse datetime string from database"""
    if not date_string:
        raise ValueError('Invalid date string')
    
    # Handle both ISO format and SQLite datetime format
    if 'T' not in date_string:
        date_string = date_string.replace(' ', 'T')
    
    try:
        return datetime.fromisoformat(date_string.replace('Z', ''))
    except ValueError:
        raise ValueError(f'Invalid date format: {date_string}')

def add_minutes_to_datetime(date_string: str, minutes: int) -> str:
    """Add minutes to a datetime string"""
    dt = parse_datetime_from_db(date_string)
    new_dt = dt + timedelta(minutes=minutes)
    return format_datetime_for_db(new_dt)

def calculate_duration_minutes(start_datetime: str, end_datetime: str) -> int:
    """Calculate duration between two datetimes in minutes"""
    try:
        start = parse_datetime_from_db(start_datetime)
        end = parse_datetime_from_db(end_datetime)
        return int((end - start).total_seconds() / 60)
    except (ValueError, TypeError):
        return 0

def is_today(date_string: str) -> bool:
    """Check if date is today"""
    try:
        date = parse_datetime_from_db(date_string)
        return date.date() == datetime.now().date()
    except (ValueError, TypeError):
        return False

def get_cooldown_end(minutes: int = 30) -> str:
    """Get cooldown end time (current time + minutes)"""
    future_time = datetime.now() + timedelta(minutes=minutes)
    return format_datetime_for_db(future_time)
